Init framewerk

tiller1010 opened the pull request:
https://werkbotstudios.teamwork.com/#/tasks/27867151

### Summary
- Restores dist related items
- Exports framewerk functions from new index.js file
- Allows certain properties to be entered with JS using CSS vars

### Testing Steps
- [x] run grunt
- [x] comment sass imports in your bundle, only import the theme file so `getThemeProperty` calls still work in your site sass

		// @import 'framewerk/sass/fw-bundle';
		// @import 'components/button/button';
		@import 'framewerk/sass/variables/theme';

- [x] `import framewerk from 'framewerk';` in your app.js
- [x] configure the framewerk with JS, change settings around

		framewerk({
			primaryColor: 'yellow',
			backgroundColorDark: 'magenta',
			componentButtonProperties: {
				backgroundColor: 'teal',
				textColor: 'blue',
				backgroundHoverColor: 'red',
				textHoverColor: 'grey',
				padding: '40px',
				margin: '50px',
				borderRadius: '25px',
				textTransform: 'uppercase',
				fontWeight: '900',
				fontFamily: 'cursive',
				fontSize: '12px'
			}
		})

- [x] make sure both your sass and js are compiled (styles will be loaded in by js, the sass should not provide any fw styles)
- [x] clear your browser cache, create a button if you need one
- [x] confirm the js fw properties are working correctly

### Issues/Concerns
- `darken` and `lighten` sass functions do not work with css vars. [`hsl` might help](https://kevingimbel.de/blog/2021/06/darken-and-lighten-colors-with-css/), but I could not get this working
- Sass math related function do not work with css vars (layoutSpace is not customizable yet for this reason)
- Cannot use css variables for media query breakpoints. Using[ css env variables](https://developer.mozilla.org/en-US/docs/Web/CSS/env) might work, but it looks like this would have to be [added to the build process](https://github.com/csstools/postcss-plugins/tree/main/plugins/postcss-env-function) (not executed at runtime with the config object)
- Certain properties may be difficult to implement this way (button variants comes to mind).
- There is a lot we could do/figure out here. This does not cover every component, and some of the above concerns may be resolved in some way. I have this set to merge into a "3" branch. I think if we break this up into a few PRs, we could handle things one-by-one until this is ready to be released.

aletail reviewed (changes requested):
Lets get this to 100% then revisit

tiller1010 commented:
@aletail This has been updated to recursively set css variables for component properties.

Button variants should also work by injecting css, and setting a scoped css variable for each property (e.g. `background-color: var(--componentButtonProperties_variants_green-btn_background-color, green) !important;`). Try with something like the following.

	const framewerk = require('werkbot-framewerk');
	// import framewerk from 'werkbot-framewerk';
	framewerk({
		componentButtonProperties: {
			variants: {
				'green-btn': {
					'background-color': 'green',
					color: 'black',
					hoverStyle: {
						'background-color': 'darkgrey',
						color: 'green',
					}
				}
			}
		}
	})

tiller1010 commented:
@aletail 

I tested this with the werkbot site. I added this test branch here, feel free to checkout the changes and adjust the settings passed into the framewerk function https://github.com/werkbot/werkbot/commit/79d544108cb2a84516fd1512f071b86fbc50e2f7

I made subtasks for each of my concerns here https://werkbotstudios.teamwork.com/#/tasks/27867151

These changes should cover https://werkbotstudios.teamwork.com/#/tasks/28059798. You should be able to include the framewerk with or without the styles loaded by js (the werkbot site is currently set to load the styles in with js).

- [ ] test the css file `composer vendor-expose`, uncomment the requirement for the css file in the PageController
- [ ] change the [last argument of the framewerk function](https://github.com/werkbot/werkbot/commit/79d544108cb2a84516fd1512f071b86fbc50e2f7#diff-d48a146c34b8a80829d9e0888bed66eecfb83861bba353b7361497367d1e953aR167) to false to exclude the styles
- [ ] change the settings around to confirm the properties passed are working

tiller1010 commented:
These recent commits should cover
https://werkbotstudios.teamwork.com/#/tasks/28059801
https://werkbotstudios.teamwork.com/#/tasks/28059802

- `calc()` is a css function, not sass, and it should work with css variables
- `@include font-size()` will now define css variables:

		elementH1Properties: {
			fontSizes: {
				desktop: {
					fontSize: '1em',
					lineHeight: '.4em',
				},
				phone: {
					fontSize: '10em',
					lineHeight: '4em',
				}
			}
		},

tiller1010 commented:
https://werkbotstudios.teamwork.com/#/tasks/28059800

The postcss env loader works for breakpoints, but the values still need to be built into the css (not injected at runtime with js). To change these breakpoint values, you wouldn't need a sass setup, but you would need to configure the [postcss env plugin](https://github.com/csstools/postcss-plugins/tree/main/plugins/postcss-env-function#usage) in your build, so there isn't much benefit to this part. You aren't forced to use sass, but I imagine more people would opt for sass instead of this.

aletail commented on README.md:
```diff
@@ -7,6 +7,28 @@ A framework of css and javascript that Werkbot uses as a foundation to build web
+### Use distributed files
+
+#### Import JS and Styles together
+`require('framewerk'); // node_modules/framewerk/dist/framewerk.js`
```
This needs update to werkbot-framewerk

aletail commented on README.md:
````diff
@@ -7,6 +7,28 @@ A framework of css and javascript that Werkbot uses as a foundation to build web
+
+#### Import JS and Styles together with initialized settings
+```
+const framewerk = require('framewerk');
````
This needs update to werkbot-framewerk

aletail reviewed (changes requested):
This looks good! 

Couple updates in the readme and just a few thoughts:
1. Does this need to be exposed? `"node_modules/werkbot-framewerk/dist"` (I suppose it does if they want to load the dist files separately?)
2. The sass functions not working like lighten and darken is not ideal (probably others?)
3. What sort of project do you think this setup approach appeals too?

My biggest concern here is our maintenance costs moving forward, 
- it seems we bend over backwards to get the sass to work using this approach (even then its not 100%) 
- Making purely sass updates - do we have to test both configurations? Might have to if we rely on sass functions like lighten and darken etc...

What issues will pop up in the future if this route is used? That is difficult to answer. I see potential issues with sass not working, dependency issues as we update/maintain the repo.

Merging this doesn't really hurt anything, but it does add some complexity to the repo.

We may want to discuss this a bit further, I see some benefit here but I also see a bit more work as well. Let me know your thoughts!!

tiller1010 commented:
> 1. Does this need to be exposed? `"node_modules/werkbot-framewerk/dist"` (I suppose it does if they want to load the dist files separately?)

Yes, if we wanted to use this css file, it would need to be public in some way. Exposing works well for a silverstripe site.

> 2. The sass functions not working like lighten and darken is not ideal (probably others?)

Yeah I think this is the biggest hang-up. [Relative colors have a specification that is is not supported yet](https://stackoverflow.com/questions/55329996/how-to-create-color-shades-using-css-variables-similar-to-darken-of-sass/55330103#55330103), maybe this will help.

> 3. What sort of project do you think this setup approach appeals too?

- Projects without a sass setup
- Apps that have a dynamic ui, and may need to initialize the js event listeners multiple times
- React apps that are not server rendered, therefore need to [initialize the js later](https://github.com/tiller1010/mern-app/blob/master/js/main.jsx#L24) (next js should not have this issue, since it has automatic ssr)
- It works well with the [next js components](https://github.com/werkbot/nextjs-install/blob/main/components/FWTextField.jsx#L14) I mentioned. Here, I pass styles unique to this field in the form of a config object. https://github.com/werkbot/nextjs-install/blob/main/components/AjaxSearchForm.jsx#L64

> We may want to discuss this a bit further, I see some benefit here but I also see a bit more work as well. Let me know your thoughts!!

I don't think this is quite ready to be merged yet. Breakpoints and color functions are the main blocks, each of which might be solved by a new css spec. I'll come back to this, but feel free to try this with next / other sites :)

tiller1010 commented:
@aletail 
I rebased this on `develop`, and I rebased the `entry-level-website` branch on the base install (which uses this).
https://github.com/werkbot/silverstripe-install/tree/entry-level-website

I renamed the task to "Release Framewerk 3.0". We could merge this, and hold off creating an actual release until we decide that this has all of the features we want and that this works well.

aletail reviewed (approved):
I tested this on the entry level branch and it worked well!

> I renamed the task to "Release Framewerk 3.0". We could merge this, and hold off creating an actual release until we decide that this has all of the features we want and that this works well.

I think that makes sense!
